test/puppetlabs/http/client/async_plaintext_test.clj

```diff
@@ -247,3 +247,109 @@
+    (with-open [client (-> (ClientOptions.)
+                           (.setConnectTimeoutMilliseconds 250)
+                           (Async/createClient))]
+      (let [request-options     (RequestOptions. "http://127.0.0.255:65535")
+            time-before-connect (System/currentTimeMillis)]
+        (is (connect-exception-thrown? (-> client
+                                           (.get request-options)
+                                           (.deref)
+                                           (.getError)))
+            "Unexpected result for connection attempt")
+        (is (elapsed-within-range? time-before-connect 2000)
+            "Connection attempt took significantly longer than timeout")))))
+
+(deftest short-connect-timeout-persistent-clojure-test-async
+  (testing (str "connection times out properly for clojure persistent client "
+                "async request with short timeout")
+    (with-open [client (async/create-client
+                         {:connect-timeout-milliseconds 250})]
+      (let [time-before-connect (System/currentTimeMillis)]
+        (is (connect-exception-thrown? (-> @(common/get
+                                              client
+                                              "http://127.0.0.255:65535")
+                                           :error))
+            "Unexpected result for connection attempt")
+        (is (elapsed-within-range? time-before-connect 2000)
+            "Connection attempt took significantly longer than timeout")))))
+
+(deftest longer-connect-timeout-test-async
+  (testing "connection succeeds for async request with longer connect timeout"
+    (testlogging/with-test-logging
+      (testwebserver/with-test-webserver app port
+        (let [url (str "http://localhost:" port "/hello")]
+          (testing "java persistent async client"
+            (with-open [client (-> (ClientOptions.)
+                                   (.setConnectTimeoutMilliseconds 2000)
+                                   (Async/createClient))]
+              (let [response (-> client
+                                 (.get (RequestOptions. url))
+                                 (.deref))]
+                (is (= 200 (.getStatus response)))
+                (is (= "Hello, World!" (slurp (.getBody response)))))))
+          (testing "clojure persistent async client"
+            (with-open [client (async/create-client
+                                 {:connect-timeout-milliseconds 2000})]
+              (let [response @(common/get client url {:as :text})]
+                (is (= 200 (:status response)))
+                (is (= "Hello, World!" (:body response)))))))))))
+
+(deftest short-socket-timeout-persistent-java-test-async
+  (testing (str "socket read times out properly for persistent java async "
+                "request with short timeout")
+    (with-open [client (-> (ClientOptions.)
+                           (.setSocketTimeoutMilliseconds 1)
```

Comment:
Relying on this to timeout in a unit test just seems really weird to me, but I'm with @cprice404, we should get this through for now and maybe discuss how this new functionality is tested later.